src/mongo/db/pipeline/expression.cpp

```diff
@@ -2716,31 +2731,110 @@ Value ExpressionIndexOfArray::evaluate(const Document& root) const {
-        Value startIndexArg = vpOperand[2]->evaluate(root);
+ExpressionIndexOfArray::Arguments ExpressionIndexOfArray::evaluateAndValidateArguments(const Document& root,
+                                                const ExpressionVector& operands,
+                                                size_t arrayLength) const {
+    
+    int startIndex = 0;
+    if (operands.size() > 2) {
+        Value startIndexArg = operands[2]->evaluate(root);
         uassertIfNotIntegralAndNonNegative(startIndexArg, getOpName(), "starting index");
-        startIndex = static_cast<size_t>(startIndexArg.coerceToInt());
+        startIndex = static_cast<int>(startIndexArg.coerceToInt());
     }
+    
 
-    size_t endIndex = array.size();
-    if (vpOperand.size() > 3) {
-        Value endIndexArg = vpOperand[3]->evaluate(root);
+    int endIndex = arrayLength;
+    if (operands.size() > 3) {
+        Value endIndexArg = operands[3]->evaluate(root);
         uassertIfNotIntegralAndNonNegative(endIndexArg, getOpName(), "ending index");
         // Don't let 'endIndex' exceed the length of the array.
-        endIndex = std::min(array.size(), static_cast<size_t>(endIndexArg.coerceToInt()));
+        endIndex = std::min(arrayLength, static_cast<size_t>(endIndexArg.coerceToInt()));
     }
 
-    for (size_t i = startIndex; i < endIndex; i++) {
-        if (getExpressionContext()->getValueComparator().evaluate(array[i] == searchItem)) {
-            return Value(static_cast<int>(i));
+    return ExpressionIndexOfArray::Arguments(vpOperand[1]->evaluate(root), startIndex, endIndex);
+}
+/**
+ * This class handles the case where IndexOfArray is given an ExpressionConstant
+ * instead of using a vector and searching through it we can use a unordered_map
+ * for O(1) lookup time.
+ */
+class ExpressionIndexOfArray::Optimized : public ExpressionIndexOfArray {
+public:
+    Optimized(const boost::intrusive_ptr<ExpressionContext>& expCtx,
+              const ValueUnorderedMap<vector<int>>& indexMap,
+              const ExpressionVector& operands)
+        : ExpressionIndexOfArray(expCtx), _indexMap(std::move(indexMap)) {
+        vpOperand = operands;
+    }
+
+    virtual Value evaluate(const Document& root) const {
+        auto args = evaluateAndValidateArguments(root, vpOperand, _indexMap.size());
+        auto indexVec = _indexMap.find(args.targetOfSearch);
+        if (indexVec != _indexMap.end()) {
+            // Search through the vector of indecies for first index in our range.
+            for (auto index : indexVec->second) {
+                if (index >= args.startIndex && index < args.endIndex) {
+                    return Value(index);
+                }
+            }
+            // the value we are searching for exists but is not in our range.
+            return Value(-1);
+        } else {
+            return Value(-1);
         }
     }
 
-    return Value(-1);
+private:
+    const ValueUnorderedMap<vector<int>> _indexMap;
```

Comment:
Can you add a comment here like the following?
```
// Maps the values in the array to the positions at which they occur. We need to remember the positions 
// so that we can verify they are in the appropriate range.
```